std/range/package.d

```diff
@@ -5887,284 +5887,387 @@ pure @safe nothrow unittest
+    {
+        return IotaResult!(B, E)(begin, end);
+    }
+    else static if (is(typeof(begin + 1)))
+    {
+        return iota(begin, end, 1);
+    }
+    else
+    {
+        static assert(false, "iota does not support the types ("~B.stringof~", "~E.stringof~").");
+    }
+}
 
-    assert(step != 0 || begin == end);
+/// Ditto
+auto iota(E)(E end)
+if (is(typeof(E(0))))
+{
+    static if (is(E == Unqual!E))
+        return iota(E(0), end);
+    else
+        return iota(Unqual!E(end));
+}
 
-    static struct Result
+/*
+ * Overload for anything supporting regular arithmetic.
+ */
+struct IotaResult(B, E, S)
+if (!anySatisfy!(isFloatingPoint, B, E, S) && is(typeof(B.init + S.init)))
+{
+    static if (is(CommonType!(B, E, S)) && !is(CommonType!(B, E, S) == void))
+        alias ValueType = CommonType!(B, E, S);
+    else static if (is(typeof(B.init + E.init + S.init)))
+        alias ValueType = typeof(B.init + E.init + S.init);
+    else static if (is(typeof(B.init + S.init)))
+        alias ValueType = typeof(B.init + S.init);
+    else
+        static assert(false, "iota does not support types that cannot be added or incremented.");
+
+    ValueType begin;
+    ValueType end;
+    S step;
+
+    private auto unsigned(T)(T value)
     {
-        private Value current, last;
-        private StepType step; // by convention, 0 if range is empty
+        static if (is(Unsigned!T)) return cast(Unsigned!T) value;
+        else return value;
+    }
 
-        this(Value current, Value pastLast, StepType step)
-        {
-            if (current < pastLast && step > 0)
-            {
-                // Iterating upward
-                assert(unsigned((pastLast - current) / step) <= size_t.max);
-                // Cast below can't fail because current < pastLast
-                this.last = cast(Value) (pastLast - 1);
-                this.last -= unsigned(this.last - current) % step;
-            }
-            else if (current > pastLast && step < 0)
-            {
-                // Iterating downward
-                assert(unsigned((current - pastLast) / (0 - step)) <= size_t.max);
-                // Cast below can't fail because current > pastLast
-                this.last = cast(Value) (pastLast + 1);
-                this.last += unsigned(current - this.last) % (0 - step);
-            }
-            else
-            {
-                // Initialize an empty range
-                this.step = 0;
-                return;
-            }
-            this.step = step;
-            this.current = current;
-        }
+    this(B b, E e, S s)
+    {
+        step = 0;
+        if (s == 0) return;
+        if (b < e && s < 0) return;
+        if (b > e && s > 0) return;
 
-        @property bool empty() const { return step == 0; }
-        @property inout(Value) front() inout { assert(!empty); return current; }
-        void popFront()
+        begin = b;
+        end = e;
+        step = s;
+
+        if (begin < end && step > 0)
```

Comment:
This is redundant because this case has already been eliminated by the early returns above. `if (step > 0)` should suffice.